### PR TITLE
Enter Marks Redesign

### PR DESCRIPTION
**Pull request by hmzahmed**
This PR includes the Redesign for Enter Marks sub-module of the module 'Exams'. A high level overview of the features in the PR includes the following:

- Display Exam Cards
- Make filters to filter through exam cards
- Ability to enter marks for a single subject
- Ability to add missing student during entering of single subject marks
- Ability to enter marks in Bulk for term exams and Monthly tests
- Print Marksheet

### CODE REVIEW

**Review by mudssrali — CHANGES_REQUESTED** [review]
Thanks @hmzahmed! It was a big module to redesign and a bunch of new features that newly incorporated. I left a few questions and added comments to improve code quality and logical constructs, please do read all and make a list of common things which we should observe while working on something new or adding tweaks to existing modules.

File: web/src/components/cards/exams/GroupedExamCard.tsx [hunk 1]
```diff
@@ -62,10 +75,11 @@ export const GroupedExamCard = ({
...
 					leaveTo="opacity-0 scale-95 ">
 					<div className="w-full">
-						{Object.entries(exams).map(([examId, exam]) => (
+						{Object.entries(exams).map(([index, exam]) => (
 							<ExamCard
-								key={examId}
-								examId={examId}
+								goToMarks={goToMarks}
+								key={exam.id}
+								examId={exam.id}
```
[the remark this answers is not in the document]
Comment (hmzahmed) [reply web/src/components/cards/exams/GroupedExamCard.tsx:82]: I was sending an array and not an object, so this was the bug that was causing undefined error.

File: web/src/pages/exams/marks/EnterBulkMarks.tsx [hunk 2]
```diff
@@ -0,0 +1,630 @@
...
+function blankExam() {
+	return {
+		id: v4(),
+		name: '',
+		subject: '',
+		total_score: '',
+		date: new Date().getTime(),
+		student_marks: {}
+	}
+}
```
[the remark this answers is not in the document]
Comment (hmzahmed) [reply web/src/pages/exams/marks/EnterBulkMarks.tsx:66]: No

File: web/src/pages/exams/marks/EnterBulkMarks.tsx [hunk 3]
```diff
@@ -0,0 +1,630 @@
...
+		const exams = this.props.exams
+		const classes = this.props.classes
+
+		const sections = getSectionsFromClasses(classes).sort(
+			(a, b) => (a.classYear || 0) - (b.classYear || 0)
+		)
+
+		const pq = queryString.parse(this.props.location.search)
+
+		const section_id = pq.section_id ? pq.section_id.toString() : ''
```
[the remark this answers is not in the document]
Comment (hmzahmed) [reply web/src/pages/exams/marks/EnterBulkMarks.tsx:83]: Its parsed query, I dont know why we are casting toString, It was the defualt logic

File: web/src/pages/exams/marks/EnterBulkMarks.tsx [hunk 4]
```diff
@@ -0,0 +1,630 @@
...
+
+	examFilter = (): ExamFilter => {
+		const pq = queryString.parse(this.props.location.search)
+
+		const exam_title = pq.exam_title ? pq.exam_title.toString() : ''
+		const year = pq.year ? pq.year.toString() : moment().format('YYYY')
+		const month = pq.month ? pq.month.toString() : ''
+
+		return {
+			exam_title,
```
[the remark this answers is not in the document]
Comment (hmzahmed) [reply web/src/pages/exams/marks/EnterBulkMarks.tsx:113]: It is defined in typings in this way and is used in many files

File: web/src/pages/exams/marks/EnterBulkMarks.tsx [hunk 5]
```diff
@@ -0,0 +1,630 @@
...
+
+		// setTimeout(() => {
+		// 	this.setState({ banner: { active: false } })
+		// }, 3000)
+	}
+
+	deleteExam = (exam_id: string): void => {
+		if (!window.confirm('Are you sure you want to delete the exam?')) {
+			return
+		}
```
[the remark this answers is not in the document]
Comment (mudssrali) [reply web/src/pages/exams/marks/EnterBulkMarks.tsx:387]: We can do this later.

**Pull request merged**